Refactor ecosystems

michaelkedar opened the pull request:
Refactor `ecosystems.py` to split each ecosystem helper implementation into its own file, which also includes the version string parser if needed.

Closes #997

michaelkedar commented:
Waiting for #996 to be merged so I can re-add its changes

michaelkedar commented:
Also thought about moving the functionality of `purl_helpers.py` into the Ecosystem helper class to avoid the duplicated ecosystem maps.

andrewpollock commented:
A preliminary readability comment while I take a closer look:
Bundling a bunch of ecosystems under semver feels unintuitive to an unfamiliar reader. If I'm digging around going "where the hell is the Go ecosystem?", it's not immediately obvious that it's buried in the `semver.py` file, because semver doesn't jump out as an ecosystem, it's a versioning scheme.

Could we have a guaranteed stub-per-ecosystem and move `semver.py` out of `osv/ecosystems` into perhaps something like `osv/versioning`?

andrewpollock commented on osv/ecosystems/_ecosystems.py:
```diff
@@ -0,0 +1,70 @@
+}
+
+
+def get(name: str) -> Ecosystem:
```
Suggestion: render this file unnecessary by moving `get()` and `normalize()` into `helper_base.py` and the rest of it into `__init__py` (which would address some of my discoverability/readability concerns)

michaelkedar replied:
I don't think I can move `get()` into `helper_base` without creating circular imports (`get` needs the `ecosystems_` dictionary, which needs to import the ecosystems implementations, which needs to import the base class)

I could move everything from `_ecosystems.py` into `__init__.py`, although I personally feel that makes it less discoverable (to me, it's not always obvious that `__init__.py` would contain actual code).

andrewpollock replied:
Ah, fair enough. Never mind then.

andrewpollock commented on osv/ecosystems/semver.py:
```diff
@@ -0,0 +1,56 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""SemVer ecosystem helper."""
```
Minor nit: Maybe reword to something like "Helper for ecosystems using SemVer"?

michaelkedar replied:
Changed to "Ecosystem helper for ecosystems using SemVer."

andrewpollock reviewed (commented):
This LGTM on the first pass, modulo my thoughts on readability/discoverability of specific SemVer-based ecosystems. How confident are you that you haven't inadvertantly removed anything with the refactoring? I haven't done a line-by-line side-by-side examination, and I'm not sure if our original test coverage was rigorous enough to pick up on this (and all the tests have changed as well).

michaelkedar commented:
> This LGTM on the first pass, modulo my thoughts on readability/discoverability of specific SemVer-based ecosystems. How confident are you that you haven't inadvertantly removed anything with the refactoring? I haven't done a line-by-line side-by-side examination, and I'm not sure if our original test coverage was rigorous enough to pick up on this (and all the tests have changed as well).

I'm pretty confident nothing (in use) was removed. I basically cut-and-pasted classes/functions from the `ecosystems.py` file into new files until it was basically empty. The tests should be the same, just moved similarly.

VSCode's not showing me any yellow squiggly lines for undefined functions/variables so I don't I accidentally deleted anything :slightly_smiling_face: 

The changes that could have potentially made an impact were:
1. The versioning implementations were moved/renamed, e.g. `osv.nuget` -> `osv.ecosystems.nuget`. I did check and `ecosystems.py` was the only file that used these.
2. The configuration variables `work_dir`, `shared_cache`, etc. had to be moved into `osv.ecosystems.config`. I searched the code base for references to these and changed them.

Other than those, importing `osv.ecosystems` should import everything that was there before. I did a search for `ecosystems` in the code base and every reference was covered.

michaelkedar commented:
> A preliminary readability comment while I take a closer look: Bundling a bunch of ecosystems under semver feels unintuitive to an unfamiliar reader. If I'm digging around going "where the hell is the Go ecosystem?", it's not immediately obvious that it's buried in the `semver.py` file, because semver doesn't jump out as an ecosystem, it's a versioning scheme.
> 
> Could we have a guaranteed stub-per-ecosystem and move `semver.py` out of `osv/ecosystems` into perhaps something like `osv/versioning`?

I've tentatively made the stub files for the SemVer-based ecosystems (though I dislike having individual files for 1 line of code). I don't really want to move `semver.py`, since it is an implementation of `Ecosystem` defined in `helper_base.py`. Perhaps it's clearer to rename it to `semver_ecosystem.py`?

A few thoughts regarding this I've had:
- The change makes `SEMVER_ECOSYSTEMS` in `_ecosystems.py` even less obvious to maintain. It might be a good idea to generate the list with `isinstance(x, SemverEcosystem)` 
- I wonder if it'd be clearer to remove the aliasing and just have the `_ecosystems` dictionary point those ecosystems to `SemverEcosystem` directly:
```python
_ecosystems = {
    'crates.io': SemverEcosystem(),
    'Go': SemverEcosystem(),
    'Hex': SemverEcosystem(),
    'Maven': Maven(),
    'npm': SemverEcosystem(),
    'NuGet': NuGet(),
    # ...
}
```

andrewpollock commented:
> Perhaps it's clearer to rename it to `semver_ecosystem.py`?

Or even throw "helper" in the name perhaps?

> The change makes SEMVER_ECOSYSTEMS in _ecosystems.py even less obvious to maintain. It might be a good idea to generate the list with isinstance(x, SemverEcosystem)

Yeah, you could do that. Is the concern that when standing up a new SemverEcosystem-based ecosystem, it gets omitted from `SEMVER_ECOSYSTEMS` in `_ecosystems.py`? There's a few ways to mitigate that. A test seems like the most obvious guardrail to me... You can also add some comments to the existing instances, signposting the existence of `SEMVER_ECOSYSTEMS` in `_ecosystems.py`

In Google3, we have the most excellent http://go/ifttt, I wonder if there's an Open Source equivalent?

> I wonder if it'd be clearer to remove the aliasing and just have the _ecosystems dictionary point those ecosystems to SemverEcosystem directly

Would that have the added benefit of GitHub doing any handy dandy symbol linkage/lookups to help people reading the code?

andrewpollock reviewed (approved):
This LGTM as is, if you wish to tweak things further per our most recent comments, go for it.

oliverchang commented on osv/ecosystems/config.py:
```diff
@@ -0,0 +1,32 @@
+
+from ..cache import Cache
+
+TIMEOUT = 30  # Timeout for HTTP(S) requests
```
Can we rename this `timeout` to be consistent with the other variables here.

oliverchang commented on osv/ecosystems/debian.py:
```diff
@@ -0,0 +1,138 @@
+    'https://storage.googleapis.com/debian-osv/first_package_output/'
+    '{version}.json')
+CACHE_DURATION_SECONDS = 60 * 60 * 24
+TIMEOUT = 30  # Timeout for HTTP(S) requests
```
Let's just use config.timeout instead?

michaelkedar replied:
Huh, I didn't notice that - sure.
I didn't introduce that duplication, I'm guessing it was added to avoid circular imports.

oliverchang reviewed (approved):
amazing work, thanks Michael!!!!

michaelkedar commented:
> You can also add some comments to the existing instances, signposting the existence of `SEMVER_ECOSYSTEMS` in `_ecosystems.py`

> > I wonder if it'd be clearer to remove the aliasing and just have the _ecosystems dictionary point those ecosystems to SemverEcosystem directly
> 
> Would that have the added benefit of GitHub doing any handy dandy symbol linkage/lookups to help people reading the code?

GitHub seems smart enough to do the symbol link either way.

I'm just going to go ahead and delete `Go`, `NPM`, etc. in favour of using `SemverEcosystem` directly, and just signpost `SEMVER_ECOSYSTEMS` with comments for now.

Maybe we can look at it again in the future if another SemVer-based ecosystem is added.
